src/multiplicity/mod.rs

```diff
@@ -0,0 +1,824 @@
+        let new_exprs: Vec<(&str, Expr)> = field_aggexpr_map
+            .into_iter()
+            .map(|(name, expr)| match &expr {
+                Expr::Aggregate(agg) => match agg.aggregate() {
+                    aggregate::Aggregate::Count => (
+                        name,
+                        Expr::multiply(
+                            Expr::col("_count_correction_factor_"),
+                            Expr::multiply(Expr::val(input.1), Expr::col(name)),
+                        ),
+                    ),
+                    aggregate::Aggregate::Sum => {
+                        (name, Expr::multiply(Expr::val(input.1), Expr::col(name)))
+                    }
+                    aggregate::Aggregate::Mean => (
+                        name,
+                        Expr::divide(Expr::col(name), Expr::col("_count_correction_factor_")),
+                    ),
+                    aggregate::Aggregate::First | aggregate::Aggregate::Last => {
+                        (name, Expr::col(name))
+                    }
+                    // panic for aggregation function that we don't know how to correct.
+                    _ => panic!(),
+                },
+                _ => (name, Expr::col(name)),
+            })
+            .collect();
+
+        let new_map: Relation = Relation::map()
+            .with((LINE_WEIGHT, Expr::col(LINE_WEIGHT)))
+            .with_iter(new_exprs.into_iter())
+            .input(reduce_with_correction)
+            .build();
+
+        RelationWithMultiplicity(new_map, 1.0)
+    }
+
+    /// propagate table weight as table weight left * table weight right
+    /// propagate LINE_WEIGHT as _LEFT_LINE_WEIGHT_ * _RIGHT_LINE_WEIGHT_
+    fn join(
+        &self,
+        join: &'a Join,
+        left: RelationWithMultiplicity,
+        right: RelationWithMultiplicity,
+    ) -> RelationWithMultiplicity {
+        let RelationWithMultiplicity(left, left_weight) = &left;
+        let RelationWithMultiplicity(right, right_weight) = &right;
+
+        let left_new_name = left.name().to_string();
+        let right_new_name = right.name().to_string();
+        let schema_names: Vec<String> =
+            join.schema().iter().map(|f| f.name().to_string()).collect();
+
+        let mut left_names = vec![format!("_LEFT{LINE_WEIGHT}")];
+        left_names.extend(schema_names.iter().take(join.left.schema().len()).cloned());
+
+        let mut right_names = vec![format!("_RIGHT{LINE_WEIGHT}")];
+        right_names.extend(schema_names.iter().skip(join.left.schema().len()).cloned());
+
+        // map old columns names (from the join) into new column names from the left and right
+        let columns_mapping: Hierarchy<Identifier> = join
+            .left
+            .schema()
+            .iter()
+            // skip 1 because the left (coming from the RelationWithMultiplicity)
+            // has the LINE_WEIGHT colum
+            .zip(left.schema().iter().skip(1))
```

Comment:
This is very fragile / error prone / hard to read.
Can you make it more readable?

Reply:
Following what has been done in the protected module I create a constant `pub const PROPAGATED_COLUMNS: usize = 1;` to be used here.